core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1513,359 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
  */
 
 static TEE_Result get_fat_start_address(uint32_t *addr);
+static TEE_Result rpmb_fs_setup(void);
 
-#if (TRACE_LEVEL >= TRACE_FLOW)
-static void dump_fat(void)
+/**
+ * fat_entry_dir_free: Free the FAT entry dir.
+ * Do not call directly but only within fat_entry_dir_deinit,
+ * or when fat_entry_dir_init fails.
+ */
+static void fat_entry_dir_free(void)
+{
+	if (fat_entry_dir->rpmb_fat_entry_buf)
+		free(fat_entry_dir->rpmb_fat_entry_buf);
+	free(fat_entry_dir);
+	fat_entry_dir = NULL;
+}
+
+/**
+ * fat_entry_dir_init: Initialize the FAT FS entry buffer/cache
+ * This function must be called before reading FAT FS entries using the
+ * function fat_entry_dir_get_next. This initializes the buffer/cache
+ * with the first FAT FS entries, i.e., 'CFG_RPMB_FS_RD_ENTRIES' when
+ * caching is disabled and
+ * MIN(CFG_RPMB_FS_CACHE_SIZE, CFG_RPMB_FS_RD_ENTRIES) when caching
+ * is enabled. After calling this function, traversing the FAT FS entries
+ * with fat_entry_dir_get_next reads from the buffer/cache and in case
+ * the end of the buffer/cache is reached, more elements are read in from RPMB.
+ * Returns TEE_SUCCESS when buffer/cache initialization succeeds.
+ */
+static TEE_Result fat_entry_dir_init(void)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	struct rpmb_fat_entry *fat_entries = NULL;
+	struct rpmb_fat_entry *fe = NULL;
 	uint32_t fat_address;
-	size_t size;
-	int i;
-	bool last_entry_found = false;
+	uint32_t num_elems_read;
```

Comment:
`uint32_t num_elems_read = 0;`
By the way, could you also init `fat_address` at line above.